### PR TITLE
Fix updating products failure when too many parameters are sent

### LINKED ISSUE #13713: Updating products fails when too many parameters are sent
## Error in Open Food Network (Rails)

**ActionController::BadRequest** in **POST /admin/products/bulk_update**
Invalid request parameters: total number of query parameters (4290) exceeds limit (4096)

[View on Bugsnag](https://app.bugsnag.com/yaycode/open-food-network-rails/errors/690da37fcc7f080b9e4908a1?event_id=69163752015408930df00000&i=gh&m=ci)

## Stacktrace

    gems/rack-2.2.20/lib/rack/query_parser.rb:194 - check_query_string

[View full stacktrace](https://app.bugsnag.com/yaycode/open-food-network-rails/errors/690da37fcc7f080b9e4908a1?event_id=69163752015408930df00000&i=gh&m=ci)

*Created by Maikel Maikel via Bugsnag*

### PR DESCRIPTION
**Pull request by chahmedejaz**
## What? Why?

- Closes #13713

This PR fixes bulk product updates failing with:

> `Invalid request parameters: total number of query parameters exceeds limit (4096)`

The root cause was that the bulk update form submitted all product fields on the page, including unchanged records and nested variant attributes. When many products were displayed, the generated payload exceeded Rails/Rack's parameter limit and resulted in a `ActionController::BadRequest` error.

### Solution

This change reduces the size of submitted bulk update payloads by submitting only changed fields and the minimum identity fields required for updates:

- Disable unchanged form fields before submission in the bulk update form.
- Preserve required product and variant identity fields (`id`) so updates can still be applied correctly.
- Ensure related hidden/derived fields (for example unit values) remain included when associated fields change.
- Normalize submitted bulk parameters server-side to support sparse submissions.
- Infer `product_id` from submitted variant IDs when a product ID is omitted (for variant-only changes).

Additionally, this PR fixes an edge case in variant unit scale updates where changing the scale (e.g. `g → kg`) without modifying the visible quantity could lead to incorrect persisted values. The update flow now normalizes `unit_value` to preserve the visible quantity while still respecting explicitly submitted values.

Tests were added to cover:
- Submitting only changed fields in bulk updates.
- Variant-only updates with inferred product IDs.
- Variant unit scale conversions and validation edge cases.

## What should we test?

- Visit the bulk product edit page (`/admin/products`).
- Load a large product list that previously produced the parameter limit error.
- Update a single product field and save.
- Confirm the update succeeds without `ActionController::BadRequest`.
- Update only a variant field (without changing product-level fields) and confirm changes persist.
- Change a variant unit scale (for example `g → kg`) without modifying the quantity and confirm the visible quantity remains consistent after save.
- Change both variant scale and quantity and confirm the explicitly entered quantity is respected.
- Verify unchanged products/variants are not unintentionally modified.
- Verify changing multiple products/variants on a single save and they all should be persisted after the save.
- Also verify creating new variants

## Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [x] Technical changes only
- [ ] Feature toggled

### CODE REVIEW

File: app/webpacker/controllers/bulk_form_controller.js [hunk 1]
```diff
@@ -148,6 +149,75 @@ export default class BulkFormController extends Controller {
...
+
+    recordContainers.forEach((container) => {
+      const elements = Array.from(container.querySelectorAll("input, select, textarea, button"));
+      const changedElements = elements.filter((element) => this.#isChanged(element));
+
+      elements.forEach((element) => {
+        if (element.type !== "submit") {
+          element.disabled = true;
+        }
+      });
```
Comment (rioug) [inline app/webpacker/controllers/bulk_form_controller.js:177]: ~~Shouldn't we be doing this after we checked if any elements have been changed ? No point disabling element if we are not going to submit anything~~
Sorry I misunderstood what we are doing here, I see we are disabling element so that they don't get submitted. I am assuming element will get re enabled after the form reload.

File: app/controllers/admin/products_v3_controller.rb [hunk 2]
```diff
@@ -334,6 +329,31 @@ def products_bulk_params
...
+
+        normalized[key] = product
+      end
+    end
+
+    def infer_product_id_from_variants(variants_attributes)
+      variant_ids = Array(variants_attributes).filter_map { |variant| variant[:id].presence }
+      return if variant_ids.empty?
+
+      Spree::Variant.where(id: variant_ids).pick(:product_id)
```
Comment (rioug) [inline app/controllers/admin/products_v3_controller.rb]: Is this to cover an edge case ? or is it likely we'll get lots of product is missing ? I fear this could generate a lots of queries and possibly result in performance issue.

**Review by rioug — APPROVED** [review]
Looks good, thanks. 
I am not super thrilled about the added complexity. I wonder, given we already have a frontend check for changed item, if we could keep a list of updated products/variants when they are changed, and use that list to create the data to send. 
But that probably comes with another set of complexity, so who knows 🤷 

Makes me wonder if we should have gone with one request per changed product/variants instead of using product sets...

File: app/webpacker/controllers/bulk_form_controller.js (hunk 1 above)
Comment (chahmedejaz) [reply app/webpacker/controllers/bulk_form_controller.js:177]: Yes, Gaetan. This will only get triggered upon form submission and after that we are reloading the entire table that would reset the all elements state to the original one.

File: app/controllers/admin/products_v3_controller.rb (hunk 2 above)
Comment (chahmedejaz) [reply app/controllers/admin/products_v3_controller.rb]: In our case it shouldn't happen because I've made sure that we always include product and variant id in the payload, so the payload ids will be used rather than db query. However, just in case if the payload doesn't include product ids then we could use a db fallback.
Let me just double check this as well.

Comment (chahmedejaz) [reply app/controllers/admin/products_v3_controller.rb]: I double checked this, and yes, we are making sure that we send the product and variant id from the frontend in the request.

Comment (rioug) [reply app/controllers/admin/products_v3_controller.rb]: Thanks for checking !

Comment (dacook) [reply app/controllers/admin/products_v3_controller.rb]: If we are making sure that we always send the product ID with the Javascript, then maybe we don't need `infer_product_id_from_variants` anymore. Can it be removed?

File: app/webpacker/controllers/bulk_form_controller.js [hunk 3]
```diff
@@ -148,6 +149,75 @@ export default class BulkFormController extends Controller {
...
+  #enableSiblingHiddenFields(element) {
+    const fieldContainer = element.closest(".field");
+    const tableCell = element.closest("td");
+    const relatedContainer = fieldContainer || tableCell;
+    if (!relatedContainer) return;
+
+    relatedContainer
+      .querySelectorAll('input[type="hidden"], input[type="checkbox"], select, textarea')
+      .forEach((relatedElement) => this.#enableElement(relatedElement));
+  }
```
Comment (dacook) [inline app/webpacker/controllers/bulk_form_controller.js]: I'm guessing this was required to ensure unit value updates work. 
It's quite specific and I worry that it might break in the future. Do you know if this specific case is covered by a system spec?

It makes me wonder if it would have been ok to take a simpler approach and enable the entire variant row if _any_ fields are changed. But this implementation is much more efficient in the case where you update one field on lots of products at once.

Which elements exactly is this for? if there's a Stimulus controller for these elements, maybe it could be marking them as "changed", then you wouldn't need this extra logic here.

File: app/services/sets/product_set.rb [hunk 4]
```diff
@@ -103,7 +103,19 @@ def update_variants_attributes(product, variants_attributes)
     def create_or_update_variant(product, variant_attributes)
       variant = find_model(product.variants, variant_attributes[:id])
       if variant.present?
+        # Capture persisted values before assignment so we can compare old vs new scale.
+        previous_state = {
+          variant_unit: variant.variant_unit,
+          unit_scale: variant.variant_unit_scale,
+          unit_value: variant.unit_value,
+        }
```
Comment (dacook) [inline app/services/sets/product_set.rb]: I think Rails does this for you: https://api.rubyonrails.org/classes/ActiveModel/Dirty.html
For example `variant.variant_unit_was`

File: app/services/sets/product_set.rb [hunk 5]
```diff
@@ -119,6 +131,62 @@ def create_or_update_variant(product, variant_attributes)
       variant&.errors.blank?
     end
 
+    def normalize_unit_value_for_scale_change(
```
Comment (dacook) [inline app/services/sets/product_set.rb]: I actually think it might be better to handle this on the browser side. It seems this code is intended to copy the behaviour on the browser, which automatically adjusts the unit value and displays it to the user. 

The problem is that the hidden field `unit_value` changes, but doesn't get marked as changed because it's a hidden field. 
We can't actually track changes on a hidden field but could work around by changing it to a text field with `display:none`
(see https://github.com/openfoodfoundation/openfoodnetwork/blob/63255e6d162f80111d0cc07529a1ce3ca0381c4e/app/webpacker/controllers/bulk_form_controller.js#L176-L178)

In that case, I think this problem would be solved. I just tried and it looks like it will work.
I added a commit to share, can you please check if this solves the cases you are aware of?

**Review by dacook — CHANGES_REQUESTED** [review]
Thanks Ahmed, I tried it out and it works beautifully. I updated 27 products and the payload was 1.6kb 👍 

I did notice that the disabled fields have a grey border around them temporarily. I think the experience would be smoother if these didn't appear, can you hide them?

Also for the second part, I have an alternative idea, can you please check it out and if it looks good, update the code and specs?

**Comment by chahmedejaz** [comment]
Hi @dacook - I've addressed your feedback and it's ready for review. Thanks

File: app/webpacker/controllers/variant_controller.js [hunk 6]
```diff
@@ -73,7 +73,14 @@ export default class VariantController extends Controller {
       unit_value = isNaN(unit_value) ? null : unit_value;
       unit_value *= this.variantUnitScale.value ? this.variantUnitScale.value : 1; // Normalise to default scale
 
-      this.unitValue.value = unit_value;
+      // Preserve original formatting when computed value matches default,
+      // to avoid false positives from number format differences (e.g. "100" vs "100.0")
+      if (unit_value !== null && !isNaN(unit_value) && !isNaN(parseFloat(this.unitValue.defaultValue))
+          && parseFloat(this.unitValue.defaultValue) === unit_value) {
+        this.unitValue.value = this.unitValue.defaultValue;
```
Comment (rioug) [inline app/webpacker/controllers/variant_controller.js]: Is it not possible to fix the default value to match the calculated one or vice versa ?

Comment (chahmedejaz) [reply app/webpacker/controllers/variant_controller.js]: No, @rioug - `defaultValue` is the floating-point value persisted in the DB, whereas the computed value is derived by parsing the user's input from `unit_value_with_description`, which is an integer. Those values diverge once the user edits the field, so we can't keep them aligned. Current changes preserves the original formatting when the value hasn't actually changed, avoiding unnecessary form-dirty noise from trivial differences like "100" vs "100.0".

**Context:**
Previously, `unit_value` was a hidden field, so its changes weren't tracked by the client-side change detection. After switching it to a text field with `display: none`, those changes are now detected. As a result, we need the client-side updates to ensure its change detection and synchronization behave correctly given the difference between the input value and the persisted DB value.

Ref: https://github.com/openfoodfoundation/openfoodnetwork/pull/14293#discussion_r3300783470

Comment (rioug) [reply app/webpacker/controllers/variant_controller.js]: I see, I was just wondering if we could avoid this extra logic but given we are dealing with user input that's not an option.Thanks !

**Review by rioug — APPROVED** [review]

File: app/webpacker/controllers/bulk_form_controller.js [hunk 7]
```diff
@@ -172,7 +172,16 @@ export default class BulkFormController extends Controller {
...
+      // Rails check_box helper generates a hidden+checkbox pair sharing the
+      // same name. When only the checkbox is re-enabled, unchecking it
+      // submits nothing (unchecked checkboxes don't submit). Enable the
+      // hidden sibling so the "0" value is sent on uncheck.
+      if (element.type === "checkbox") {
+        const hiddenSibling = element.previousElementSibling;
+        if (hiddenSibling?.type === "hidden" && hiddenSibling.name === element.name) {
+          this.#enableElement(hiddenSibling);
+        }
+      }
```
Comment (dacook) [inline app/webpacker/controllers/bulk_form_controller.js:214]: Nice fix, _and_ explanation 💯

File: app/webpacker/controllers/bulk_form_controller.js [hunk 8]
```diff
@@ -233,13 +233,16 @@ export default class BulkFormController extends Controller {
...
-        opt.hasAttribute("selected"),
+      // Compare by value instead of DOM element reference.
+      // Tom Select can create duplicate <option> DOM elements during remote
+      // search, so reference comparison (===) fails after revert.
+      const defaultOption = Array.from(element.options).find((opt) =>
+        opt.defaultSelected,
       );
-      const selectedOption = element.selectedOptions[0];
-      const areBothBlank = selectedOption.value === "" && defaultSelected === undefined;
+      const defaultValue = defaultOption?.value ?? "";
```
Comment (dacook) [inline app/webpacker/controllers/bulk_form_controller.js:270]: TIL about the "Nullish Coalescing Operator". I am definitely not going to remember that name, but hopefully I'll remember the code!

File: app/webpacker/controllers/bulk_form_controller.js [hunk 9]
```diff
@@ -233,13 +233,16 @@ export default class BulkFormController extends Controller {
...
+      const noDefaultSelected = defaultOption === undefined && element.value === "";
 
-      return !areBothBlank && selectedOption !== defaultSelected;
+      return !noDefaultSelected && element.value !== defaultValue;
```
Comment (dacook) [inline app/webpacker/controllers/bulk_form_controller.js:273]: Can this change be detected with a JS test? Perhaps we should be moving this function into its own separate module and tested with a unit test..
`±` But until then maybe it will be easy to include the case in `spec/javascripts/stimulus/bulk_form_controller_test.js`? 
If not, then don't worry about it.

File: app/webpacker/controllers/variant_controller.js (hunk 6 above)
Comment (dacook) [reply app/webpacker/controllers/variant_controller.js]: `±` Another idea: can we avoid it by formatting the value in the ruby view? At the moment, the form helper is automatically loading the value, I guess with `to_s`. But if we add something like `, value: number_with_precision(variant.unit_value, strip_insignificant_zeros: true)` it might do the trick.

I tested this with `100.0.to_d` and it comes back as "100".

Note: I chose this method instead of the simpler `.round` method just in case sometimes we _do_ need decimal points (I can't remember if we do, but it seems wise to avoid modifying the number just in case).

If that helps simplify things here, I think it's worth a shot.

**Review by dacook — CHANGES_REQUESTED** [review]
Great, thanks for updating this Ahmed, it's looking really good. I have a couple more suggestions to update, hopefully minor changes.

**Comment by chahmedejaz** [comment]
Thanks @dacook - I've addressed your feedback here and it's ready for review.
https://github.com/openfoodfoundation/openfoodnetwork/pull/14293/changes/05355cb448cf24a61d5a75b264f59f2b655e0325

**Review by dacook — APPROVED** [review]
Great!

**Comment by AMEA-LYON** [comment]
Hi @chahmedejaz , 
I found this scenario that is generating an error : see screen capture 
It seems that pre-filled values are lost when creating a new variant...

https://github.com/user-attachments/assets/558285f5-e946-46e4-b882-28ea3e48373e

Same error message when changing unit scale : 

https://github.com/user-attachments/assets/d278de8f-fdbd-47a8-a0c9-7f260066cf3b

**Comment by chahmedejaz** [comment]
Hi @AMEA-LYON - Thanks for testing this out. Strange, I'm not able to reproduce it on my end. Let me look into it more deeply on the staging. Thanks!

**Comment by chahmedejaz** [comment]
@AMEA-LYON - I've confirmed this, it's related to some recent changes in the master that's affecting the changes in this PR. I'm taking a look at it and will get back to you with the fix. Thanks!

**Comment by chahmedejaz** [comment]
@AMEA-LYON - The above issues have been address and ready for re-testing. Thanks!

**Comment by AMEA-LYON** [comment]
Hi @chahmedejaz , 
pre-filled variants are now handled correctly, thanks ! 🎉 

But I still have the issue when changing the unit scale (exactly the same way...) 😞

**Comment by chahmedejaz** [comment]
@AMEA-LYON - I'm not able to reproduce this on my machine. Seems like the issue occurs when the unit scale is downsized. I tried that on my machine but still it's working fine. Let me investigate the issue further on the UK env. Thanks for sharing.

**Comment by AMEA-LYON** [comment]
Another remark : if I enter a new price with 2 decimal digits (ex 1.23), it's correctly recorded,
but if I enter 3 decimal digits (ex 1.234) the price turns to 1,234.00 )

**Comment by chahmedejaz** [comment]
Hi @AMEA-LYON, I've debugged these issues and it looks like they're not related to this PR. I'm able to reproduce them on the current release version (`v5.7.5`).

Could you please verify whether you can also reproduce them on the current `master` branch and the current release version? If the issues exist there as well, I think it would be best to create separate bug issues so they can be tracked independently of this PR.

Regarding the first issue, where downsizing the unit scale throws an error, I suspect it may be related to the data on the UK instance. Could you also try reproducing it on a different instance to confirm whether it's data-specific?

Thanks!

**Comment by AMEA-LYON** [comment]
Hi @chahmedejaz ,

Concerning the price issue, I reproduced them also on v5.7.5 and master, so it is not related to this PR, I will create a new issue.

Downsizing the unit scale is working on Staging AU, good point ! 🎉 Ready to go !
